### DISCUSSION TITLE
Improve distribution

### DISCUSSION THREAD

**Discussion by Borewit** [Borewit/listFix#162] [Ideas]
Can you try to run this distribution?:
@Touwys [listFix-2.7.0-9 build on windows-2019 build agent](https://github.com/Borewit/listFix/suites/12166720707/artifacts/642131092), which is the build result from this [run](https://github.com/Borewit/listFix/actions/runs/4671358528#summary-12671505965).

You need to unzip the zip, and unzip zip you find in that zip. Then run the `.bat` file.

**Comment by touwys** [comment]
@Borewit 

> Can you try to run this distribution?

**_Result_: listFix-2.7.0-9 build on windows-2019 build agent**

After running the bat-file, uninstructed, I opted to run the jar-file as indicated in the annotated Screenshot, provided herewith:

• 
**[Screenshot](https://user-images.githubusercontent.com/65730626/231435991-1e316d5e-8641-4b7b-ba2a-e74db4530b73.png)**
•

**Comment by Borewit** [comment]
I thought this was compiled, but I should have better look, this is still Java.
So it still requires at least a JRE 15 installed.

**Reply by touwys** [reply]
> I thought this was compiled, but I should have better look, this is still Java. So it still requires at least a JRE 15 installed.

@Borewit, When you say "compiled" — do you mean compiled for running on Windows 7?

**Reply by Borewit** [reply]
> @Borewit, When you say "compiled" — do you mean compiled for running on Windows 7?

Good question @touwys.

What I tried to do, is make a distribution of listFix() to what [jlink](https://docs.oracle.com/en/java/javase/17/docs/specs/man/jlink.html) calls a "_custom runtime image_", which a vague description in my view. 

Good old fashion Java code you compile and converts Java source code into classes. Classes are byte code which can be executed by the JRE present on the OS. 

[jlink](https://docs.oracle.com/en/java/javase/17/docs/specs/man/jlink.html) turns Java into a _custom runtime image_ which does not require a JRE. This image includes the JRE. I would like to explore this option.

**Comment by touwys** [comment]
@Borewit:

Is your main focus now on completing the development of a self-contained Java JDK distribution — before you re-engage with the current range of issues, improvements, etc? Or, is its development concurrent with the listFix() release we've so far been testing? In other words, is developing the envisaged new distribution going to take precedence over fixing the issues and improvements over the release we've been working on? Alternatively, are the two going to be run in parallel?

Can you please give us some insight into what you have planned next?

**Comment by Borewit** [comment]
Can you try this distribution build with jlink: https://workupload.com/file/e6fdNAshfP6
No local JRE/JDK required, that is build in.
`image/bin/listtfix.bat` is what you need to run to execute it.

It's quite a lot of work this distribution. In a way I would like to prioritize that. All-tough I have not much expectation, I like make the application more attractive to a bigger audience. Ideally get to some more involvement in the project and have the feeling it is actually being used by reasonable community (more then just you and me @touwys ;-).  

For the same reason I like to make another release soon.

**Reply by touwys** [reply]
> Can you try this distribution build with jlink...

👍🏻 


> It's quite a lot of work this distribution. In a way I would like to prioritize that. All-tough I have not much expectation, I like make the application more attractive to a bigger audience. Ideally get to some more involvement in the project and have the feeling it is actually being used by reasonable community (more then just you and me @touwys ;-).

You are 100% correct. I often think about that too. With such a small audience, and little reward, your dedication is admirable. May your audience grows strongly. However, looking at it realistically, the music scene has changed so drastically in recent years that the demand for tools, such as listFix(), may have diminished. Or, maybe, other people do not use, or break, playlists as often as we do? 😁

**Reply by touwys** [reply]
> Can you try this distribution build with jlink: https://workupload.com/file/e6fdNAshfP6

I extracted the image to a folder on my main drive — as indicated in the screenshot below — and then proceeded to run the yellow-highlighted Java-executable. You did not specify which executable I was to run after having extracted the image, therefore I am uncertain whether I chose the correct one. There are two Java-executables present in the same folder.

_The result of running this particular executable, was a momentary flash of the old DOS-window across the screen — nothing else beyond that._

•
**[Screenshot of the bin-folder](https://user-images.githubusercontent.com/65730626/232990149-2500772e-476f-4c18-82ec-d63d61f377a3.png)**
•

**Reply by Borewit** [reply]
> No local JRE/JDK required, that is build in.
`image/bin/listtfix.bat` is what you need to run to execute it.

**Comment by touwys** [comment]
> No local JRE/JDK required, that is build in.

Yes, that part was clear. I also did run the bat-file as required. It is what should happen next that I am not at all clear about. What should I expect to see after having run the bat-file? Could you please explain the next step, or steps, I have take to get listFix() to start?

**Reply by Borewit** [reply]
There is no next step my friend, that should start listFix(). To bad that doesn't work for you. Any error message if you run it in a DOS box?

**Comment by touwys** [comment]
> There is no next step my friend, that should start listFix()

OK, thanks for clearing that up.

> Any error message if you run it in a DOS box?

The DOS box that opened after executing the bat-file, was visible for a split-second only — the duration too short to read anything. And, that was the only visible sign that something did in fact happen, nothing else. Should I try it in DOS using other means?

**Reply by Borewit** [reply]
If you run `cmd` and run the `listfix.bat` from there, you will be able to read the error message.

**Comment by touwys** [comment]
My apologies for the delay in reporting back, @Borewit. Access to the PC is still severely hampered by the endless scheduled electricity blackouts. This abnormality is severely impacting on society at large. 

I took another shot at running the bat-file, this time **successfully**, using the DOS-console of my go-to file-manager, xplorer2. It was indeed also possible to start listFix() straight away this time, by just clicking on the bat-file. Could the Windows security software possibly have interfered, previously? 

Being in a hurry to pass the DOS error-log results on to you, I otherwise tested listFix() only _superficially._ After the app start-up, the initial preparations, such as adding the media and playlist directories, proceeded normally. I also test-repaired two playlists, and then restarted the app before stopping the test. The only small error I noticed straight away was the listFix() version number "null", as per screenshot.

🏆 **Well done, @Borewit!** 

What's up next?

_____

_**Screenshots and log:**_

•
[xplorer2 DOS-Console](https://user-images.githubusercontent.com/65730626/233262848-e383bf96-6b31-46bc-bfd3-7d8cdfe3b3f1.png)
•
**[DOS-Error Log: listfix_jl_test_win7_20230420.txt](https://github.com/Borewit/listFix/files/11279941/listfix_jl_test_win7_20230420.txt)**
•
[DOS-Error Image](https://user-images.githubusercontent.com/65730626/233261437-4eeaf46d-38d7-4d5d-b518-fbef336b47b0.png)
• 
[Version Number "null"](https://user-images.githubusercontent.com/65730626/233266209-91390094-ba8c-49d1-997f-b93053418e74.png)
•

_____

**Comment by Borewit** [comment]
That is great news @touwys, it looks like we got better foundation to build a distribution.
 Yes, security software can kick and prevent you from running all sorts of things, although the default Windows security is not very intrusive.

1. I need to turn these changes in a proper PR. _Update: #164_ 
2. Also need to finalize the required changes to [Lizzy](https://github.com/Borewit/lizzy/pull/67), ideally 
3. Optionally create a PR for [dd-plist to update to migrate to JPMS](https://github.com/3breadt/dd-plist/issues/85) which Lizzy depends on
4. Make the runtime-image as artifact available on from the pipeline (so no manual build and upload required form my end). And you are guaranteed to be able to pull a constistent build from each PR.
5. Turn this "runtime image" into a distribution, with proper icon etc 
6. Try to build the distribution on the pipe-line, so it automatically available after each change. 

You can prepare [#120](https://github.com/Borewit/listFix/issues/120#issuecomment-1516809821) if you like... (will continue writing in that one)

**Comment by touwys** [comment]
> Yes, security software can kick and prevent you from running all sorts of things, although the default Windows security is not very intrusive.

In addition to Windows Defender, there is Malwarebytes that could have been, more likely, the source of such trouble, but I did not find any sign of it interfering. Anyway, after now having repeated successful app restarts, this is no longer of any importance.

There is one hindrance though:  The persistent presence of the DOS-dialogue. It remains open, even after listFix() is already restarted. This dialogue should close automatically, and _the error logged to file_. I suppose this will taken care of, during further development?

> You can prepare #120 if you like...

See my comments there — at #120.


----------------------------------------

20 Apr 2023 21:00:23 Borewit ***@***.***>:

**Comment by touwys** [comment]
@Borewit 

**Re: Deep Playlist Repairs**:

_Does the Deep playlist repair serve any further purpose after the removal of the "Find Exact Matches" repair?_

•••

![image](https://user-images.githubusercontent.com/65730626/236669064-f079d38d-5a1e-4340-8c4d-7ac7f6fa34ed.png)

•••

**Reply by Borewit** [reply]
I guess the added value is that it allows a user to run the "closest match repair" on multiple selected playlists.

**Reply by touwys** [reply]
> _Does the Deep playlist repair serve any further purpose after the removal of the "Find Exact Matches" repair?_

I still think that, unless there are compelling reasons against it, we should stick to only avenue of playlist repair, and to develop that to its pinnacle.

•••

**Reply by Borewit** [reply]
We can rename "Deep" to "Repair".

**Reply by touwys** [reply]
> I guess the added value is that it allows a user to run the "closest match repair" on multiple selected playlists.

But, isn't that already attainable through "Repair all open playlists"?

**Reply by touwys** [reply]
> We can rename "Deep" to "Repair".

Do you perhaps use "Deep" repair often? I never have, and do not see me ever using it. It just overcomplicates the matching. Keep it straightforward?

**Comment by Borewit** [comment]
I don't think it is the same, but it is similar.

The repair open playlists, repairs all playlists open in editors.

The "Deep" button allows repairing via a playlists-directories playlist selection.

I use the application rarely 🤣, so I am not sure if my usage statistics add much value.

**Comment by touwys** [comment]
> The "Deep" button allows repairing via a playlists-directories playlist selection.

Quite so, but that's exactly my point: it overcomplicates things, i.e. "similar", but not the "same". Two more or less identical ways to accomplish the same goal. Note that we've now already made it much easier to simply drag and drop playlists into the Editor, quickly and efficiently, from where "Repair _all open_ playlists" is run. Why, then, keep "Deep" repair as another, now redundant avenue to access playlist on which to carry out similar repairs? It does not make sense?

> I use the application rarely

😮 I agree, it is not an app that finds use every single day. It is, however, a keeper, in my books. It's one of a kind, and one never knows when next it will be needed.

**Reply by Borewit** [reply]
For the time being I started with renaming that button, as this is in line with the path we already took in PR #157 in https://github.com/Borewit/listFix/commit/a08bd5ef249bc9e0b48255b0335325279e0c91c4 (direct commit on main branch)

**Comment by Borewit** [comment]
Released [v2.8.0](https://github.com/Borewit/listFix/releases/tag/v2.8.0) !

**Reply by touwys** [reply]
> Released [v2.8.0](https://github.com/Borewit/listFix/releases/tag/v2.8.0) !

@Borewit: Great! Installed it already, and first impressions are all good.

**Comment by pgraffman** [comment]
Is there any documentation on how to install and run this on windows?  Java is installed and I run the listFix.2.2.0.exe and get an error message "You need to add a media directory before you can find the new locations of your files.  See help for more information." There is no help files.

**Reply by Borewit** [reply]
Installation of  Java is **no** prerequisite.  

- As @touwys recommended,  install latest release: https://github.com/Borewit/listFix/releases
- Before you can fix a playlist, the media directory (library) where your audio files are stored. When fixing playlists, only audio files within your media library are assumed to exist.

**Comment by touwys** [comment]
> I run the listFix.2.2.0.exe and get an error message "You need to add a media directory before you can find the new locations of your files.

@pgraffman

The most recent version of the app is: [**listFix() version 2.10.0**](https://github.com/Borewit/listFix/releases/tag/v2.10.0)

...

The message you are seeing is not an error message, but a programme instruction. Allow the app to open normally after seeing this message. For listFix() to work properly from the outset, and beyond, it is imperative that you add your preferred Media Directories. Otherwise, listFix() is rendered incapable of performing its primary task of locating the closest matches for the tracks that require repair. Similarly, you will have to add those playlists which you wish to repair, to Playlist Directories pane. (You may also add playlists to the Playlist Editor pane, if you so wish.)

[**Screenshot**](https://github.com/Borewit/listFix/assets/65730626/e295a726-dfe3-42d5-8356-631fabd51740)

**Comment by pgraffman** [comment]
Got it. The new release fixed the situation,

 

Thank you.

 

 

Peter Graffman

 

Office: 781-444-8233

Mobile: 617-680-6662

FAX: 781-465-6187
